ra/ra.go

```diff
@@ -1795,8 +1819,327 @@ func (ra *RegistrationAuthorityImpl) revokeCertificate(ctx context.Context, cert
+// for the given certificate.
+// TODO(#5152) make the issuerID argument an issuance.IssuerNameID
+func (ra *RegistrationAuthorityImpl) purgeOCSPCache(ctx context.Context, cert *x509.Certificate, issuerID int64) error {
+	issuer, ok := ra.issuersByNameID[issuance.IssuerNameID(issuerID)]
+	if !ok {
+		// TODO(#5152): Remove this fallback (which only gets used when revoking by
+		// serial, so the issuer ID had to be read from the db).
+		issuer, ok = ra.issuersByID[issuance.IssuerID(issuerID)]
+		if !ok {
+			return fmt.Errorf("unable to identify issuer of cert with serial %q", core.SerialToString(cert.SerialNumber))
+		}
+	}
+
+	purgeURLs, err := akamai.GeneratePurgeURLs(cert, issuer.Certificate)
+	if err != nil {
+		return err
+	}
+
+	_, err = ra.purger.Purge(ctx, &akamaipb.PurgeRequest{Urls: purgeURLs})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RevokeCertByApplicant revokes the certificate in question. It allows any
+// revocation reason from (0, 1, 3, 4, 5, 9), because Subscribers are allowed to
+// request any revocation reason for their own certificates. However, if the
+// requesting RegID is an account which has authorizations for all names in the
+// cert but is *not* the original subscriber, it overrides the revocation reason
+// to be 5 (cessationOfOperation), because that code is used to cover instances
+// where "the certificate subscriber no longer owns the domain names in the
+// certificate". It does not add the key to the blocked keys list, even if
+// reason 1 (keyCompromise) is requested, as it does not demonstrate said
+// compromise. It attempts to purge the certificate from the Akamai cache, but
+// it does not hard-fail if doing so is not successful, because the cache will
+// drop the old OCSP response in less than 24 hours anyway.
+func (ra *RegistrationAuthorityImpl) RevokeCertByApplicant(ctx context.Context, req *rapb.RevokeCertByApplicantRequest) (*emptypb.Empty, error) {
+	if req == nil || req.Cert == nil || req.RegID == 0 {
+		return nil, errIncompleteGRPCRequest
+	}
+
+	if _, present := revocation.UserAllowedReasons[revocation.Reason(req.Code)]; !present {
+		return nil, berrors.BadRevocationReasonError(req.Code)
+	}
+	if !features.Enabled(features.MozRevocationReasons) {
+		// By our current policy, demonstrating key compromise is the only way to
+		// get a certificate revoked with reason key compromise. Upcoming Mozilla
+		// policy may require us to allow the original Subscriber to assert the
+		// keyCompromise revocation reason, even without demonstrating such.
+		if req.Code == ocsp.KeyCompromise {
+			return nil, berrors.BadRevocationReasonError(req.Code)
+		}
+	}
+
+	cert, err := x509.ParseCertificate(req.Cert)
+	if err != nil {
+		return nil, err
+	}
+
+	serialString := core.SerialToString(cert.SerialNumber)
+
+	logEvent := certificateRevocationEvent{
+		ID:           core.NewToken(),
+		SerialNumber: serialString,
+		Reason:       req.Code,
+		Method:       "applicant",
```

Comment:
Can you confirm my understanding: `"method": "applicant"` will only ever occur if we get an error from GetSerialMetadata; otherwise we'll log one of `"method": "subscriber"` or `"method": "control"`?